sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -1319,3 +1319,123 @@ case class BRound(child: Expression, scale: Expression)
 }
+
+/**
+ *  Returns the bucket number into which
+ *  the value of this expression would fall after being evaluated.
+ *
+ * @param expr is the expression for which the histogram is being created
+ * @param minValue is an expression that resolves
+ *                 to the minimum end point of the acceptable range for expr
+ * @param maxValue is an expression that resolves
+ *                 to the maximum end point of the acceptable range for expr
+ * @param numBucket is an expression that resolves to
+ *                  a constant indicating the number of buckets
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(expr, min_value, max_value, num_bucket) - Returns the `bucket` to which operand would be assigned in an equidepth histogram with `num_bucket` buckets, in the range `min_value` to `max_value`.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_(5.35, 0.024, 10.06, 5);
+       3
+  """)
+// scalastyle:on line.size.limit
+case class WidthBucket(
+  expr: Expression,
+  minValue: Expression,
+  maxValue: Expression,
+  numBucket: Expression) extends Expression with ImplicitCastInputTypes {
+
+  override def children: Seq[Expression] = Seq(expr, minValue, maxValue, numBucket)
+  override def foldable: Boolean = children.drop(1).forall(_.foldable)
+  override def inputTypes: Seq[AbstractDataType] = Seq(DoubleType, DoubleType, DoubleType, LongType)
+  override def dataType: DataType = LongType
+  override def nullable: Boolean = true
+
+  private lazy val _minValue: Any = minValue.eval()
+  private lazy val minValueV = _minValue.asInstanceOf[Double]
+
+  private lazy val _maxValue: Any = maxValue.eval()
+  private lazy val maxValueV = _maxValue.asInstanceOf[Double]
+
+  private lazy val _numBucket: Any = numBucket.eval()
+  private lazy val numBucketV = _numBucket.asInstanceOf[Long]
+
+  private val errMsg = "The argument [%d] of WIDTH_BUCKET function is NULL or invalid."
+
+  override def eval(input: InternalRow): Any = {
+
+    if (foldable) {
+      if (_minValue == null) {
+        throw new RuntimeException(errMsg.format(2))
+      } else if (_maxValue == null) {
+        throw new RuntimeException(errMsg.format(3))
+      } else if (_numBucket == null || numBucketV <= 0) {
+        throw new RuntimeException(errMsg.format(4))
+      } else {
+        val exprV = expr.eval(input)
+        if (exprV == null) {
+          null
+        } else {
+          MathUtils.widthBucket(exprV.asInstanceOf[Double], minValueV, maxValueV, numBucketV)
+        }
+      }
+    } else {
+      val evals = children.map(_.eval(input))
+      val invalid = evals.zipWithIndex.filter { case (e, i) =>
+        (i > 0 && e == null) || (i == 3 && e.asInstanceOf[Long] <= 0)
+      }
+      if (invalid.nonEmpty) {
+        invalid.foreach(l => throw new RuntimeException(errMsg.format(l._2 + 1)))
+      } else if (evals(0) == null) {
+        null
+      } else {
+        MathUtils.widthBucket(
```

Comment:
`MathUtils.widthBucket` is used only for this new expr now? If so, could we move the method to this expr?